chain/tests/test_pibd_validation.rs

```diff
@@ -0,0 +1,235 @@
+use crate::chain::types::NoopAdapter;
+use crate::core::core::pmmr;
+use crate::core::core::{hash::Hashed, pmmr::segment::SegmentIdentifier};
+use crate::core::{genesis, global, pow};
+
+use croaring::Bitmap;
+
+mod chain_test_helper;
+
+fn test_pibd_chain_validation_impl(is_test_chain: bool, src_root_dir: &str) {
+	global::set_local_chain_type(global::ChainTypes::Mainnet);
+	let mut genesis = genesis::genesis_main();
+	// Height at which to read kernel segments (lower than thresholds defined in spec - for testing)
+	let mut target_segment_height = 11;
+
+	if is_test_chain {
+		global::set_local_chain_type(global::ChainTypes::AutomatedTesting);
+		genesis = pow::mine_genesis_block().unwrap();
+		target_segment_height = 3;
+	}
+
+	{
+		println!("Reading Chain, genesis block: {}", genesis.hash());
+		let dummy_adapter = Arc::new(NoopAdapter {});
+
+		// The original chain we're reading from
+		let src_chain = Arc::new(
+			chain::Chain::init(
+				src_root_dir.into(),
+				dummy_adapter.clone(),
+				genesis.clone(),
+				pow::verify_size,
+				false,
+			)
+			.unwrap(),
+		);
+
+		// For test compaction purposes
+		/*src_chain.compact().unwrap();
+		src_chain
+		.validate(true)
+		.expect("Source chain validation failed, stop");*/
+
+		let sh = src_chain.get_header_by_height(0).unwrap();
+		println!("Source Genesis - {}", sh.hash());
+
+		let horizon_header = src_chain.txhashset_archive_header().unwrap();
+
+		println!("Horizon header: {:?}", horizon_header);
+
+		// Copy the header from source to output
+		// Not necessary for this test, we're just validating the source
+		/*for h in 1..=horizon_height {
+			let h = src_chain.get_header_by_height(h).unwrap();
+			dest_chain.process_block_header(&h, options).unwrap();
+		}*/
+
+		// Init segmenter, (note this still has to be lazy init somewhere on a peer)
+		// This is going to use the same block as horizon_header
+		let segmenter = src_chain.segmenter().unwrap();
+
+		// BITMAP - Read + Validate, Also recreate bitmap accumulator for target tx hash set
+		// Predict number of leaves (chunks) in the bitmap MMR from the number of outputs
+		let bitmap_mmr_num_leaves =
+			(pmmr::n_leaves(horizon_header.output_mmr_size) as f64 / 1024f64).ceil() as u64;
+		println!("BITMAP PMMR NUM_LEAVES: {}", bitmap_mmr_num_leaves);
+
+		// And total size of the bitmap PMMR
+		let bitmap_pmmr_size = pmmr::peaks(bitmap_mmr_num_leaves + 1)
+			.last()
+			.unwrap_or(&pmmr::insertion_to_pmmr_index(bitmap_mmr_num_leaves))
+			.clone();
+		println!("BITMAP PMMR SIZE: {}", bitmap_pmmr_size);
+		println!(
+			"Bitmap Segments required: {}",
+			SegmentIdentifier::count_segments_required(bitmap_pmmr_size, target_segment_height)
+		);
+		// TODO: This can probably be derived from the PMMR we'll eventually be building
+		// (check if total size is equal to total size at horizon header)
+		let identifier_iter =
+			SegmentIdentifier::traversal_iter(bitmap_pmmr_size, target_segment_height);
```

Comment:
Since you just computed count_segments_required, you might as well check that it equals the identifier_iter size?!

Reply:
Yes, I was doing that at one point but removed it (led to redundant code since `count()` on an iterator consumes it, I'm sure there are other ways but didn't go further). Will add something similar into future tests.